### PR TITLE
Don't remove commands from cache when skipped

### LINKED ISSUE #14213: Timers of banned instances are forever re-triggered
**Describe the bug**

During [the game day](https://confluence.camunda.com/display/ZEEBE/2023-09-07+Game+Day), we had a case of a loop around a timer catch event which ended up producing an enormous amount of records. To recover, we thought we could just ban the instances. Turns out, even when banned, the timers kept re-triggering, and the system was writing timer trigger commands over and over, since it's not aware of banned instances.

> **Note**
> We were also confused by the ban metrics, since banning the same process instance twice will increment the counter twice, even if really, only one instance was banned.

**To Reproduce**

Deploy this beautiful looping process:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" xmlns:modeler="http://camunda.org/schema/modeler/1.0" xmlns:camunda="http://camunda.org/schema/1.0/bpmn" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Web Modeler" exporterVersion="54e0234" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="8.2.0" camunda:diagramRelationId="eb53fcb1-5305-43cf-8c22-9b5c514460cf">
  <bpmn:process id="GamedayProcess" isExecutable="true">
    <bpmn:startEvent id="StartEvent_1" name="Send Email newsletter">
      <bpmn:outgoing>Flow_0yg4sg4</bpmn:outgoing>
    </bpmn:startEvent>
    <bpmn:sequenceFlow id="Flow_0yg4sg4" sourceRef="StartEvent_1" targetRef="Event_02vb3b2" />
    <bpmn:intermediateCatchEvent id="Event_02vb3b2" name="16 o&#39;clock">
      <bpmn:incoming>Flow_0yg4sg4</bpmn:incoming>
      <bpmn:incoming>Flow_0ucihdu</bpmn:incoming>
      <bpmn:incoming>Flow_1dr09o3</bpmn:incoming>
      <bpmn:outgoing>Flow_0ucihdu</bpmn:outgoing>
      <bpmn:outgoing>Flow_0bbnbt4</bpmn:outgoing>
      <bpmn:timerEventDefinition id="TimerEventDefinition_1mcgxr7">
        <bpmn:timeDuration xsi:type="bpmn:tFormalExpression">=date and time(today(), time("09:00:00@Europe/Berlin")) - now()</bpmn:timeDuration>
      </bpmn:timerEventDefinition>
    </bpmn:intermediateCatchEvent>
    <bpmn:sequenceFlow id="Flow_1537k1v" sourceRef="Gateway_1knikyw" targetRef="Activity_1r2d7uw" />
    <bpmn:endEvent id="Event_1rb64gr">
      <bpmn:incoming>Flow_1jfq4dj</bpmn:incoming>
    </bpmn:endEvent>
    <bpmn:parallelGateway id="Gateway_1knikyw">
      <bpmn:incoming>Flow_0bbnbt4</bpmn:incoming>
      <bpmn:outgoing>Flow_1dr09o3</bpmn:outgoing>
      <bpmn:outgoing>Flow_1537k1v</bpmn:outgoing>
    </bpmn:parallelGateway>
    <bpmn:sequenceFlow id="Flow_1jfq4dj" sourceRef="Activity_1r2d7uw" targetRef="Event_1rb64gr" />
    <bpmn:serviceTask id="Activity_1r2d7uw" name="Send Email">
      <bpmn:extensionElements>
        <zeebe:taskDefinition type="sendNewsletter" />
      </bpmn:extensionElements>
      <bpmn:incoming>Flow_1537k1v</bpmn:incoming>
      <bpmn:outgoing>Flow_1jfq4dj</bpmn:outgoing>
    </bpmn:serviceTask>
    <bpmn:sequenceFlow id="Flow_0ucihdu" sourceRef="Event_02vb3b2" targetRef="Event_02vb3b2" />
    <bpmn:sequenceFlow id="Flow_1dr09o3" name="Next Day again" sourceRef="Gateway_1knikyw" targetRef="Event_02vb3b2" />
    <bpmn:sequenceFlow id="Flow_0bbnbt4" sourceRef="Event_02vb3b2" targetRef="Gateway_1knikyw" />
  </bpmn:process>
  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="GamedayProcess">
      <bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="StartEvent_1">
        <dc:Bounds x="152" y="120" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="141" y="166" width="57" height="27" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_121xx9h_di" bpmnElement="Event_02vb3b2">
        <dc:Bounds x="252" y="120" width="36" height="36" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="255" y="93" width="50" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Event_1rb64gr_di" bpmnElement="Event_1rb64gr">
        <dc:Bounds x="572" y="120" width="36" height="36" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Gateway_0i1s8cx_di" bpmnElement="Gateway_1knikyw">
        <dc:Bounds x="335" y="113" width="50" height="50" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape id="Activity_16cmsco_di" bpmnElement="Activity_1r2d7uw">
        <dc:Bounds x="420" y="98" width="100" height="80" />
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge id="Flow_0yg4sg4_di" bpmnElement="Flow_0yg4sg4">
        <di:waypoint x="188" y="138" />
        <di:waypoint x="252" y="138" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1537k1v_di" bpmnElement="Flow_1537k1v">
        <di:waypoint x="385" y="138" />
        <di:waypoint x="420" y="138" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1jfq4dj_di" bpmnElement="Flow_1jfq4dj">
        <di:waypoint x="520" y="138" />
        <di:waypoint x="572" y="138" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0ucihdu_di" bpmnElement="Flow_0ucihdu">
        <di:waypoint x="252" y="138" />
        <di:waypoint x="252" y="138" />
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_1dr09o3_di" bpmnElement="Flow_1dr09o3">
        <di:waypoint x="360" y="163" />
        <di:waypoint x="360" y="210" />
        <di:waypoint x="270" y="210" />
        <di:waypoint x="270" y="156" />
        <bpmndi:BPMNLabel>
          <dc:Bounds x="277" y="192" width="76" height="14" />
        </bpmndi:BPMNLabel>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNEdge id="Flow_0bbnbt4_di" bpmnElement="Flow_0bbnbt4">
        <di:waypoint x="288" y="138" />
        <di:waypoint x="335" y="138" />
      </bpmndi:BPMNEdge>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</bpmn:definitions>
```

Run it, wait for it to build some state, then ban it. The system will keep writing timer trigger commands. So our force ban feature works, but it doesn't allow us to recover from all cases.

I suspect more process-instance-adjacent commands (timers, jobs, etc.) may have other issues as well.

**Expected behavior**

Banning instance stops any activity related to that instance.

-------

https://jira.camunda.com/browse/SUPPORT-23524

### PR DESCRIPTION
**Pull request by berkaycanbc**
## Description

A boolean field is introduced to `ProcessingResult` and `ProcessingResultBuilder` to allow updating the
field. This field will be set to `true` when a command is skipped (e.g. banned instance's
commands). When a `ProcessingResult` is marked as `skipped`, we do not remove the command from the
command cache to prevent appending it to the log over and over again.

## Related issues

closes #14213

### CODE REVIEW

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/processing/TimerTriggerBannedInstanceIT.java
Comment (korthout) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/processing/TimerTriggerBannedInstanceIT.java:118]: ❌ This limit is placed wrongly causing slow test execution. 

There is no `TRIGGERED` event on the banned process instance, only on the second (non-banned) process instance. So, this limit simply waits 5 seconds and then continues. To ensure a fast test, you'll need to flip the order of these calls. First limit for a `TRIGGERED` event, then filter the stream for the records of the banned process instance, before finally asserting that only a single `TRIGGER` command exists.

File: zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/processing/TimerTriggerBannedInstanceIT.java
Comment (korthout) [inline zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/processing/TimerTriggerBannedInstanceIT.java:120]: 🔧 `.onlyCommands()` is redundant after filtering for the `TRIGGER` intent

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @berkaycanbc 

💭 I'm still in favor of options 2 and 3 over the proposal in this PR. When we take either of those options, we'd revert the changes in this PR again. So, my proposal is not to take merge this PR and instead apply option 2 or 3 immediately. AFAIK, [you already found a solution](https://github.com/camunda/camunda/issues/14213#issuecomment-2239304772) for options 2 or 3.

❌ In case you do want to continue with this PR: I've requested changes that must be applied before merging.

💭 These changes would also require a review from ZDP, as it changes the stream-platform.

**Comment by berkaycanbc** [comment]
@korthout thanks for reviewing the PR!

> 💭 I'm still in favor of options 2 and 3 over the proposal in this PR. When we take either of those options, we'd revert the changes in this PR again. So, my proposal is not to take merge this PR and instead apply option 2 or 3 immediately. AFAIK, https://github.com/camunda/camunda/issues/14213#issuecomment-2239304772 for options 2 or 3.

I agree actually, this one was a half day effort, so will not hurt to throw away. We can start with implementing option 3 and tackle the option 2 in a separate PR. If we implement them successfully, we can close this one.

**Comment by korthout** [comment]
In that case let's close this PR @berkaycanbc

**Pull request closed without merging**
